### PR TITLE
Info Bar Directive 

### PR DESCRIPTION
**Pull request by petzlux**
I have tried to copy the scaleline directive, and change it to an attribution control, to explore how the creation of custom controls works. Unfortunately, I cant seem to make it work, it doesn't give me any errors, but there is no attribution either ... 

Can someone point me in the right direction ?

### CODE REVIEW

**Comment by elemoine** [comment]
Where are the attributions supposed to go? I do not see attributions on the [mockup](http://dev.camptocamp.com/files/lux/geoportail_lux_specs/cartes/). Will they be placed in the info bar together with the scaleline?

**Comment by petzlux** [comment]
No its simply a test for me to see how adding a ol3 control through a directive works,  at the very least I would hope to see the console.log to execute,  but can't see that

File: geoportailv3/static/js/infobardirective.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    controller: 'AppInfoBarController',
+    controllerAs: 'ctrl',
+    bindToController: true,
+    template: '<div ngeo-control="ctrl.control"' +
+        'ngeo-control-map="ctrl.map">'
+  };
+};
+
+
+app.module.directive('appInfoBar', app.infobarDirective);
```
Comment (elemoine) [inline geoportailv3/static/js/infobardirective.js]: If you use `<app-infobar>` in the HTML then the name of the directive should be `appInfobar`. That's your error. Angular just ignores the DOM element, because there is no directive associated to it.

File: geoportailv3/static/js/infobardirective.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * @return {angular.Directive} The Directive Object Definition.
+ * @ngInject
+ */
+app.infobarDirective = function() {
+  return {
+    restrict: 'E',
+    scope: {
+      'map': '=appInfoBarMap'
+    },
```
Comment (elemoine) [inline geoportailv3/static/js/infobardirective.js]: Use `appInfobarMap` here.

File: geoportailv3/static/js/infobardirective.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+app.module.directive('appInfoBar', app.infobarDirective);
+
+
+
+/**
+ * @ngInject
+ * @constructor
+ */
+app.InfoBarDirectiveController = function() {
```
Comment (elemoine) [inline geoportailv3/static/js/infobardirective.js]: `InfoBarDirectiveController` -> `InfobarDirectiveController` for consistency reasons here.

File: geoportailv3/static/js/infobardirective.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                this['ctrl'].mouseposition.setProjection(projection);
+                return _coordFormat(coord, projection.getCode());
+            };
+            return _coordFormat(coord, $scope.projection.value)
+        }
+    });
+    this['map'].addControl(this.mouseposition);
+
+    function _utmZoneCheck(coord){
+        var lonlat = ol.proj.transform(coord, $scope.projection.value, "EPSG:4326");
```
Comment (pgiraud) [inline geoportailv3/static/js/infobardirective.js]: `$scope.projection.value` will equal "EPSG:3263*" which is not a valid projection.
You probably want to use "map.getView().getProjection()" instead.

Comment (petzlux) [inline geoportailv3/static/js/infobardirective.js]: see https://github.com/Geoportail-Luxembourg/geoportailv3/commit/218af1e20d85348eab20cc68bbf2d93de3122c03

its only a placeholder signifiying that we need to check which UTM band we should display  ...

**Comment by petzlux** [comment]
@pgiraud & @elemoine : 

Yesterday we tried again to add proj4js as a dependency, to allow us to transform from EPSG:32362 / 32361 to EPSG:4326  (see 89c797b )

we added proj4 as a package in package.json , and added it as a src in build.json. We also added it in main.js and infobardirective.js as a goog.require('proj4') . NPM during build installs proj4, but the compiler then always complains that he cant find the package. 

This is now blocking finalisation of mouse coordinate display functionality. Can you please assist us in resolving this issue.

Thanks

**Comment by pgiraud** [comment]
OK. I'm having a look.

**Comment by pgiraud** [comment]
Is the code where you added proj4 to the dependencies pushed somewhere? I can't see it in this pull-request (ie. info_bar branch).

**Comment by pgiraud** [comment]
So, proj4 need to be loaded in the page.
You'll need to add the following in `template.html`:
```html
<script src="${request.static_url('%s/proj4/dist/proj4-src.js' % node_modules_path)}"></script>
```

Also, you'll have to create the proj4 definition files.

```javascript
goog.provide('app.proj_2169');

goog.require('ol.Map');
goog.require('ol.proj');

proj4.defs('EPSG:2169',
    '+proj=tmerc +lat_0=49.83333333333334 +lon_0=6.166666666666667 ' +
    '+k=1 +x_0=80000 +y_0=100000 +ellps=intl ' +
    '+towgs84=-189.681,18.3463,-42.7695,-0.33746,-3.09264,2.53861,0.4598 ' +
    '+units=m +no_defs"');
ol.proj.get('EPSG:2169').setExtent([48225.17, 56225.60, 105842.04, 139616.40]);
```

Please pay attention to indentation and syntax errors in your code. It won't build in its current state.

**Comment by elemoine** [comment]
What's important to note is that proj4 should be used as an external library. It should not be compiled with the application. ol3 already includes externs definitions for proj4.

**Comment by rmichaelis** [comment]
Could someone explain me when we have to compile with the application and when we have to include as an external library ?

**Comment by petzlux** [comment]
Thanks for these helpful comments, I am implementing right now and the problem is resolved !  Will push with changes to this pull request once finalised.

**Comment by elemoine** [comment]
@rmichaelis, we compile code that can be compiled. As simple as that. In practice the app is compiled with OpenLayers 3, ngeo and Closure Library. For libraries that cannot be compiled (because their code is not compatible with Closure Compiler) we use externs definitions. The externs definitions provide types, and prevent renaming.

**Comment by rmichaelis** [comment]
Okay, I understand now. Thanks for the information.

**Comment by petzlux** [comment]
I have difficulties with the way closure compiler handles property values in https://github.com/Geoportail-Luxembourg/geoportailv3/commit/035de28ce4502b94c827fe3fa1f7bd709d6ad468

As a result, the coordinates display in the different projections only works properly in debug mode (where no variables are renamed). I have tried to convert to bracket notation but to no avail. 

Can someone @pgiraud / @elemoine point out what is going wrong here ?

**Comment by pgiraud** [comment]
In your case, you forgot to use strings when you create the `mouseposition` property:
```javascript
this['mouseposition'] = new ol.control.MousePosition(...)
```

File: geoportailv3/static/js/infobar/elevationdirective.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  return {
+    restrict: 'E',
+    scope: {
+      'map': '=appElevationMap'
+    },
+    controller: 'AppElevationController',
+    controllerAs: 'ctrl',
+    bindToController: true,
+    template: '<span class="elevation" translate>' +
+                'Elevation: {{elevation}} m</span>'
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/elevationdirective.js]: You use `controllerAs: 'ctrl'` (which is goog), so you want to use `{{ctrl.elevation}}` in the template. This requires placing `elevation` on the controller instead of directly on the scope.

File: geoportailv3/static/js/infobar/elevationdirective.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+app.module.directive('appElevation', app.elevationDirective);
+
+
+
+/**
+ * @ngInject
+ * @constructor
+ */
+app.ElevationDirectiveController = function($scope, $http, ngeoDebounce) {
+  this['map'].on('pointermove',
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/elevationdirective.js]: We don't want to trigger requests when the map is panned. So `ngeoDebounce` should only be called when `e.dragging` is false:

```js
this['map'].on('pointermove',
    /**
     * @param {ol.MapBrowserEvent} e Map browser event.
     */
    function(e) {
      if (e.dragging) {
        return;
      }
      ngeoDebounce(…);
    });
```

File: geoportailv3/static/js/infobar/elevationdirective.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+      ngeoDebounce(function(e) {
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/elevationdirective.js]: It would be good to type the callback function argument.

File: geoportailv3/static/js/infobar/elevationdirective.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+          method: 'GET',
+          params: {'lon': lonlat[0], 'lat': lonlat[1]}
+        }).
+            success(function(data) {
+              if (data['dhm'] > 0) {
+                $scope['elevation'] = parseInt(data['dhm'] / 100 , 0);
+              } else {
+                $scope['elevation'] = 'N/A';
+              }
+            });
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/elevationdirective.js]: So, to store the elevation on the controller, you can do as follows:

```js
    success(goog.bind(function(data) {
      this['elevation'] = data['dhm'] > 0 ? parseInt(data['dhm'] / 100, 0) : 'N/A';
    }, this));
```

File: geoportailv3/static/js/infobar/infobardirective.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+app.module.directive('appInfobar', app.infobarDirective);
+
+
+
+/**
+ * @ngInject
+ * @constructor
+ */
+app.InfobarDirectiveController = function($scope, $document) {
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/infobardirective.js]: You have unused arguments.

File: geoportailv3/static/js/infobar/projectionselector.html [hunk 10]
```diff
@@ -0,0 +1,6 @@
+ <select ng-model="ctrl.projection" 
+                ng-options="opt as opt.label for opt in ctrl.projectionOptions" ng-change="ctrl.switchProjection()"></select>
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselector.html]: You can use one-time binding for `ctrl.projectionOptions`.

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+                   var xhdms = hdms.split(' ').slice(4, 8).join(' ');
+                   var template = xhdms + ' ({x}) | ' + yhdms + ' ({y})';
+                   return ol.coordinate.format(coord, template, 5);
+        case 'EPSG:32632':
+                   return ol.coordinate.format(coord, '{x} | {y} (UTM32N)', 0);
+        case 'EPSG:32631':
+                   return ol.coordinate.format(coord, '{x} | {y} (UTM31N)', 0);
+    }
+};
+/**
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: You need two empty lines above this line.

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+ */
+app.ProjectionselectorDirectiveController.prototype.switchProjection = function() {
+    var projection = null;
+    if (this['projection']['value'] === 'EPSG:3263*') {
+        projection = ol.proj.get('EPSG:32632');
+    } else {
+        projection = ol.proj.get(this['projection']['value']);
+    }
+    var widget = this.mouseposition;
+    console.log(projection);
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: `console.log` to be removed.

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    widget.setProjection(projection);
+    widget.setCoordinateFormat(
+       /** @type {ol.CoordinateFormatType} */ (angular.bind(this, function(coord) {
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: We use `gooo.bind` instead of `angular.bind`, as `angular.bind` won't be renamed by the compiler. In general, we use `goog` function when available.

File: geoportailv3/templates/index.html [hunk 14]
```diff
@@ -196,7 +197,9 @@ <h2 translate>infos</h2>
     <script src="${request.route_url('deps.js')}"></script>
     <script src="${request.static_url('geoportailv3:static/js/main.js')}"></script>
     <script src="${request.static_url('%s/ngeo/utils/watchwatchers.js' % node_modules_path)}"></script>
+    <script src="${request.static_url('%s/proj4/dist/proj4-src.js' % node_modules_path)}"></script>
 % else:
+    <script src="${request.static_url('%s/proj4/dist/proj4-src.js' % node_modules_path)}"></script>
```
Comment (elemoine) [inline geoportailv3/templates/index.html]: Instead, we want to add proj4 to `build.js`. See how it's done in the Makefile.

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+ * @private
+ */
+app.ProjectionselectorDirectiveController.prototype.coordFormat_ = function(coord, epsg_code)
+{
+    if (epsg_code === 'EPSG:3263*') {
+        var projection = ol.proj.get(this.utmZoneCheck_(coord));
+        this.mouseposition.setProjection(projection);
+        epsg_code = projection.getCode();
+    }
+    console.log(epsg_code);
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: `console.log` to be removed.

**Comment by petzlux** [comment]
I managed to resolve the issues last Friday, forgot to remove the tags. 

@elemoine thanks, will take into account code comments

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+/**
+ * @private
+ */
+app.ProjectionselectorDirectiveController.prototype.utmZoneCheck_ = function(coord) {
+    var lonlat = /** @type {ol.Coordinate} */ (ol.proj.transform(coord, this.mouseposition.getProjection() , 'EPSG:4326'));
+    if (Math.floor(lonlat[0]) >= 6) {
+      return 'EPSG:32632';
+    } else {
+      return 'EPSG:32631';
+    }
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: How about:

```js
  return Math.floor(lonlat[0]) >= 6 ? 'EPSG:32632' : 'EPSG:32631';
```

?

File: geoportailv3/static/js/infobar/projectionselector.html [hunk 17]
```diff
@@ -0,0 +1,6 @@
...
+<div class="mouse-coordinates"></div>
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselector.html]: You could use the `ngeo-control` directive here.

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+  this['projection'] = this['projectionOptions'][0];
+  this.mouseposition = new ol.control.MousePosition({
+    className: 'custom-mouse-coordinates',
+    target: $document.find('.mouse-coordinates')[0],
+    projection: this.projection['value'],
+    coordinateFormat: /** @type {ol.CoordinateFormatType} */ (angular.bind(this, function(coord) {
+      return this.coordFormat_(coord, this['projection']['value']);
+    }))
+  });
+  this.map.addControl(this.mouseposition) ;
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: As indicated above you could rely on the `ngeo-control` directive. In that case you just need to make the mouse position control available in the controller:

```js
  this['mousePositionControl'] = new ol.control.MousePosition({
    // …
  });
```

The `ngeo-control` directive will take care of adding the control to the map.

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+app.ProjectionselectorDirectiveController = function($scope, $document) {
+  this['projectionOptions'] = [
+    {'label': 'LUREF', 'value': 'EPSG:2169'},
+    {'label': 'Long/Lat WGS84', 'value': 'EPSG:4326'},
+    {'label': 'WGS84 UTM 32|31', 'value': 'EPSG:3263*'}
+  ];
+  this['projection'] = this['projectionOptions'][0];
+  this.mouseposition = new ol.control.MousePosition({
+    className: 'custom-mouse-coordinates',
+    target: $document.find('.mouse-coordinates')[0],
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: The `target` option should be removed if you use `ngeo-control`, as suggested.

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+    projection: this.projection['value'],
+    coordinateFormat: /** @type {ol.CoordinateFormatType} */ (angular.bind(this, function(coord) {
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: Use `goog.bind`.

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+    }
+};
+
+/**
+ * @private
+ */
+app.ProjectionselectorDirectiveController.prototype.coordFormat_ = function(coord, epsg_code)
+{
+    if (epsg_code === 'EPSG:3263*') {
+        var projection = ol.proj.get(this.utmZoneCheck_(coord));
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: Indent is incorrect. It is strange that this is not caught by the linters.

File: geoportailv3/static/js/infobar/projectionselectordirective.js (hunk 19 above)
Comment (petzlux) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: @elemoine Can you please explain in more detail how to make use of ngeo-control in this context. The example in ngeo repo didnt really help me.

File: geoportailv3/static/js/infobar/projectionselector.html [hunk 22]
```diff
@@ -0,0 +1,6 @@
+ <select ng-model="ctrl.projection" 
+                ng-options="opt as opt.label for opt in ::ctrl.projectionOptions" ng-change="ctrl.switchProjection()"></select>
+<div class="mouse-coordinates"></div>
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselector.html]: To use `ngeo-control` you will need the following here for the "mouse-coordinates" div:

```html
<div class="mouse-coordinates" ngeo-control="ctrl.mousePositionControl" ngeo-control-map="ctrl.map"></div>
```

File: geoportailv3/static/js/infobar/projectionselectordirective.js (hunk 19 above)
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: Use the following in the partial (as indicated above):

```html
<div class="mouse-coordinates" ngeo-control="ctrl.mousePositionControl" ngeo-control-map="ctrl.map"></div>
```

and that in the controller constructor:

```js
  this['mousePositionControl'] = new ol.control.MousePosition({
    className: 'custom-mouse-coordinates',
    projection: this.projection['value'],
    coordinateFormat: …
  });
```

Note that there's no `target` anymore, and no call to `map.addControl`. That's taken care of by the ngeo-control directive.

File: geoportailv3/static/js/infobar/infobardirective.js [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * @export
+ */
+app.InfobarDirectiveController.prototype.infobarSwitch = function() {
+  if (this['infobarOpen']) {
+    this['infobarOpen'] = false;
+  } else {
+    this['infobarOpen'] = true;
+  }
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/infobardirective.js]: Can be changed to

```js
this['infobarOpen'] = !this['infobarOpen'];
```

File: geoportailv3/static/js/infobar/projectionselector.html [hunk 24]
```diff
@@ -0,0 +1,11 @@
+<div ng-class="::{'dropdown': true, 'dropup': true}">
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselector.html]: Any reason for using `ng-class` here? May the classes change?

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+    function(epsg_code) {
+  var projection = null;
+  if (epsg_code === 'EPSG:3263*') {
+    projection = ol.proj.get('EPSG:32632');
+  } else {
+    projection = ol.proj.get(epsg_code);
+  }
+  this['projection'] = goog.array.find(this['projectionOptions'],
+      function(obj) {
+        return obj.value == epsg_code;
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: You probably want to use `obj['value']` here.

File: geoportailv3/static/js/infobar/projectionselector.html (hunk 24 above)
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselector.html]: And any reason for having both `dropdown` and `dropup`?

File: geoportailv3/templates/index.html (hunk 14 above)
Comment (elemoine) [inline geoportailv3/templates/index.html]: As I said already, we want to add proj4 to `build.js`. But we can do that with a separate PR if you do not feel comfortable with it.

**Comment by elemoine** [comment]
I've added a few more comments. Otherwise it looks very good to me!

File: geoportailv3/templates/index.html (hunk 14 above)
Comment (rmichaelis) [inline geoportailv3/templates/index.html]: in issue #250 , you said that "proj4 should be used as an external library. It should not be compiled with the application" 
Why is it different now ?

Comment (elemoine) [inline geoportailv3/templates/index.html]: It is not :-)

The `build.js` file located in `geoportailv3/static/build`, and loaded by the page in non-debug mode, is a concatenation of all the js builds. More specifically it's a concatenation of `jquery.min.js`, `angular.min.js`, `angular-animate.min.js`, `angular-gettext.min.js`, `bootstrap.min.js`, and `.build/build.js`, which is the app + ol3 + ngeo + closure Closure-compiled code. And soon, it will also include the application's directive partials (https://github.com/Geoportail-Luxembourg/geoportailv3/pull/293). I hope this clarifies.

Comment (rmichaelis) [inline geoportailv3/templates/index.html]: Yes it clarifies :-) thanks

Comment (petzlux) [inline geoportailv3/templates/index.html]: Can I take you up on the offer and merge this as is, and we deal with build.js in separate pull request.

Comment (elemoine) [inline geoportailv3/templates/index.html]: Yes, that was my suggestion as well.

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+
+
+/**
+ * @private
+ * @param {Array} coord
+ * @param {string} epsg_code
+ * @return {string}
+ */
+app.ProjectionselectorDirectiveController.prototype.coordFormat_ =
+    function(coord, epsg_code) {
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: Why not use `epsgCode`?

File: CONST_requirements.txt [hunk 27]
```diff
@@ -3,5 +3,5 @@
 --find-links http://pypi.camptocamp.net/internal-pypi/index/c2cgeoportal-win
 -r CONST_versions.txt
 -e git+https://github.com/camptocamp/pyramid_closure#egg=pyramid_closure
--e git+https://github.com/Pylons/pyramid@1e02bbfc0df09259bf207112acf019c8dba44a90#egg=pyramid
+-e git+https://github.com/sbrunner/pyramid@c2cgeoportal#egg=pyramid
```
Comment (elemoine) [inline CONST_requirements.txt]: @sbrunner, do we really want to use that?

**Comment by petzlux** [comment]
Ready to merge ? @elemoine

File: geoportailv3/static/js/infobar/elevationdirective.js [hunk 28]
```diff
@@ -0,0 +1,69 @@
+/**
+ * @fileoverview This file provides a "elevation" directive. This directive is
+ * used to insert Elevation information into the HTML page.
+ * Example:
+ *
+ * <app-elevation app-elevation-map="::mainCtrl.map"map></app-elevation>
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/elevationdirective.js]: There's a typo here.

Comment (petzlux) [inline geoportailv3/static/js/infobar/elevationdirective.js]: done

**Comment by elemoine** [comment]
> Ready to merge ? @elemoine

I just noticed a typo in one of the examples. Please merge when this is fixed. And we need to create GitHub tickets for the remaining issues.

**Comment by elemoine** [comment]
@petzlux feel free to delete the `info_bar` branch :-)
